## Final Project Submission

Please fill out:
* Student name: WILLIAM NYAWIR
* Student pace: full time
* Scheduled project review date/time: 
* Instructor name: Mrs MARYANNE MWIKALE
* Blog post URL:


In [89]:
# IMPORTING LIBRARIES
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sqlite3

# DATA UNDERSTANDING

In [90]:
numbers = pd.read_csv(r"data/numbers.csv")
numbers

,id,release_date,movie,production_budget,domestic_gross,worldwide_gross
0,1,"Dec 18, 2009",Avatar,"$425,000,000","$760,507,625","$2,776,345,279"
1,2,"May 20, 2011",Pirates of the Caribbean: On Stranger Tides,"$410,600,000","$241,063,875","$1,045,663,875"
2,3,"Jun 7, 2019",Dark Phoenix,"$350,000,000","$42,762,350","$149,762,350"
3,4,"May 1, 2015",Avengers: Age of Ultron,"$330,600,000","$459,005,868","$1,403,013,963"
4,5,"Dec 15, 2017",Star Wars Ep. VIII: The Last Jedi,"$317,000,000","$620,181,382","$1,316,721,747"
...,...,...,...,...,...,...
5777,78,"Dec 31, 2018",Red 11,"$7,000",$0,$0
5778,79,"Apr 2, 1999",Following,"$6,000","$48,482","$240,495"
5779,80,"Jul 13, 2005",Return to the Land of Wonders,"$5,000","$1,338","$1,338"
5780,81,"Sep 29, 2015",A Plague So Pleasant,"$1,400",$0,$0


In [91]:
box_office=pd.read_csv(r"data/box_office_mojo.csv")
box_office

,title,studio,domestic_gross,foreign_gross,year
0,Toy Story 3,BV,415000000.0,652000000,2010
1,Alice in Wonderland (2010),BV,334200000.0,691300000,2010
2,Harry Potter and the Deathly Hallows Part 1,WB,296000000.0,664300000,2010
3,Inception,WB,292600000.0,535700000,2010
4,Shrek Forever After,P/DW,238700000.0,513900000,2010
...,...,...,...,...,...
3382,The Quake,Magn.,6200.0,NaN,2018
3383,Edward II (2018 re-release),FM,4800.0,NaN,2018
3384,El Pacto,Sony,2500.0,NaN,2018
3385,The Swan,Synergetic,2400.0,NaN,2018


In [92]:
# Connect to the SQLite database
conn = sqlite3.connect(r'data/im.db')

cur=conn.cursor()

cur.execute("""SELECT name FROM sqlite_master WHERE type = 'table';""")
table_names = cur.fetchall()
table_names


[('movie_basics',),
 ('directors',),
 ('known_for',),
 ('movie_akas',),
 ('movie_ratings',),
 ('persons',),
 ('principals',),
 ('writers',)]

# UNDERSTANDING 'NUMBERS' DATA

In [93]:
numbers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5782 entries, 0 to 5781
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   id                 5782 non-null   int64 
 1   release_date       5782 non-null   object
 2   movie              5782 non-null   object
 3   production_budget  5782 non-null   object
 4   domestic_gross     5782 non-null   object
 5   worldwide_gross    5782 non-null   object
dtypes: int64(1), object(5)
memory usage: 271.2+ KB


The above information contains general information about the numbers dataset

In [94]:
numbers.describe()

,id
count,5782.000000
mean,50.372363
std,28.821076
min,1.000000
25%,25.000000
50%,50.000000
75%,75.000000
max,100.000000


The above information contains the statistical attributes of the numbers dataset

In [95]:
numbers.shape

(5782, 6)

The numbers dataset has 5782 rows and 6 columns

In [96]:
numbers.ndim

2

The numbers dataset has a total of 2 dimensions

In [97]:
numbers.columns

Index(['id', 'release_date', 'movie', 'production_budget', 'domestic_gross',
       'worldwide_gross'],
      dtype='object')

The numbers dataset has 6 columns in total

In [98]:
numbers.isna().sum()

id                   0
release_date         0
movie                0
production_budget    0
domestic_gross       0
worldwide_gross      0
dtype: int64

The numbers dataset has 0 missing values in each column

In [99]:
numbers.duplicated().sum()

0

The numbers dataset has 0 duplcated values in each column

# UNDERSTANDING 'BOX OFFICE' DATASET

In [100]:
box_office.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3387 entries, 0 to 3386
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   title           3387 non-null   object 
 1   studio          3382 non-null   object 
 2   domestic_gross  3359 non-null   float64
 3   foreign_gross   2037 non-null   object 
 4   year            3387 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 132.4+ KB


This is the general information about box office dataset

In [101]:
box_office.describe()

,domestic_gross,year
count,3.359000e+03,3387.000000
mean,2.874585e+07,2013.958075
std,6.698250e+07,2.478141
min,1.000000e+02,2010.000000
25%,1.200000e+05,2012.000000
50%,1.400000e+06,2014.000000
75%,2.790000e+07,2016.000000
max,9.367000e+08,2018.000000


This is the statistical information of the box office dataset

In [102]:
box_office.shape

(3387, 5)

The box office dataset contains 3387 rows and 5 columns

In [103]:
box_office.isna().sum()

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

The box office dataset has missing values. The column with the highest missing values is the foreign_gross

In [104]:
box_office.duplicated().sum()  

0

The box office dataset has no duplicated values

# UNDERSTANDING 'IMDB' DATASET

In [105]:
cur.execute("""SELECT name FROM sqlite_master WHERE type = 'table';""")
table_names = cur.fetchall()
table_names

[('movie_basics',),
 ('directors',),
 ('known_for',),
 ('movie_akas',),
 ('movie_ratings',),
 ('persons',),
 ('principals',),
 ('writers',)]

Finding the tables in the IMDB dataset

In [106]:
imdb=pd.read_sql(""" SELECT movie_ratings.movie_id,genres,runtime_minutes,start_year,primary_title,
             movie_ratings.numvotes , movie_ratings.averagerating
             FROM movie_basics
            JOIN movie_ratings
            ON movie_basics.movie_id = movie_ratings.movie_id""",conn)

imdb

,movie_id,genres,runtime_minutes,start_year,primary_title,numvotes,averagerating
0,tt0063540,"Action,Crime,Drama",175.0,2013,Sunghursh,77,7.0
1,tt0066787,"Biography,Drama",114.0,2019,One Day Before the Rainy Season,43,7.2
2,tt0069049,Drama,122.0,2018,The Other Side of the Wind,4517,6.9
3,tt0069204,"Comedy,Drama",NaN,2018,Sabse Bada Sukh,13,6.1
4,tt0100275,"Comedy,Drama,Fantasy",80.0,2017,The Wandering Soap Opera,119,6.5
...,...,...,...,...,...,...,...
73851,tt9913084,Documentary,75.0,2019,Diabolik sono io,6,6.2
73852,tt9914286,"Drama,Family",98.0,2019,Sokagin Çocuklari,136,8.7
73853,tt9914642,Documentary,NaN,2017,Albatross,8,8.5
73854,tt9914942,None,NaN,2019,La vida sense la Sara Amat,5,6.6


Choosing the columns that will be used from the specified tables (movie_basics, movie_id)

In [107]:
imdb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73856 entries, 0 to 73855
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   movie_id         73856 non-null  object 
 1   genres           73052 non-null  object 
 2   runtime_minutes  66236 non-null  float64
 3   start_year       73856 non-null  int64  
 4   primary_title    73856 non-null  object 
 5   numvotes         73856 non-null  int64  
 6   averagerating    73856 non-null  float64
dtypes: float64(2), int64(2), object(3)
memory usage: 3.9+ MB


This is the general information of the IMDB dataset

In [108]:
imdb.describe()

,runtime_minutes,start_year,numvotes,averagerating
count,66236.000000,73856.000000,7.385600e+04,73856.000000
mean,94.654040,2014.276132,3.523662e+03,6.332729
std,208.574111,2.614807,3.029402e+04,1.474978
min,3.000000,2010.000000,5.000000e+00,1.000000
25%,81.000000,2012.000000,1.400000e+01,5.500000
50%,91.000000,2014.000000,4.900000e+01,6.500000
75%,104.000000,2016.000000,2.820000e+02,7.400000
max,51420.000000,2019.000000,1.841066e+06,10.000000


This is the statistical data of the imdb dataset 

In [109]:
imdb.isna().sum()

movie_id              0
genres              804
runtime_minutes    7620
start_year            0
primary_title         0
numvotes              0
averagerating         0
dtype: int64

Every column in imdb dataset has no missing values except from genres and runtime_minutes with 804 and 7620 null values respectively

In [110]:
imdb.nunique()

movie_id           73856
genres               923
runtime_minutes      289
start_year            10
primary_title      69993
numvotes            7349
averagerating         91
dtype: int64

The above information contains number of uniques values in each column in the dataset

In [111]:
imdb.duplicated(['primary_title']).sum()

3863

This is the amount of duplicated values in the primary title.

In [112]:
imdb.duplicated(['movie_id']).sum()

0

There is no duplicated values in the movie_id column

In [113]:
#length of movie_id and length of primary_title should be same as they are unique identifiers
len(imdb['movie_id']) == len(imdb['primary_title'])

True

This means that each unique movie id is assigned to a specific primary title

# DATA CLEANING

In [114]:
# Dropping Columns
# Missing value
# Duplicate value
# Unique value
# Outlier detection
# Feature engineering

## Cleaning the numbers data

The numbers dataset is clean


In [ ]:
# numbers.to_csv('cleaned_numbers_data.csv', index=False)

## Cleaning box office data

In [115]:
# Box Office Mojo Data Cleaning

box_office.isna().sum()

title                0
studio               5
domestic_gross      28
foreign_gross     1350
year                 0
dtype: int64

In [116]:
print(box_office['domestic_gross'].dtype)
d_gross_mean = int(box_office['domestic_gross'].mean())
d_gross_mean

float64


28745845

In [117]:
type(box_office['foreign_gross'])
# Finding the mean of foreign_gross column which is a pandas series
box_office['foreign_gross'] = pd.to_numeric(box_office['foreign_gross'], errors='coerce')
f_gross_mean = round(box_office['foreign_gross'].mean())
f_gross_mean

75057042

In [118]:
# Filling missing values with mean
box_office['domestic_gross'].fillna(d_gross_mean, inplace=True)
box_office['foreign_gross'].fillna(f_gross_mean, inplace=True)

#Filling missing studio values with mode
studio_mode = box_office['studio'].mode()
print(studio_mode)
box_office['studio'].fillna(studio_mode, inplace=True)

0    IFC
Name: studio, dtype: object


In [119]:
box_office.isna().sum()

title             0
studio            5
domestic_gross    0
foreign_gross     0
year              0
dtype: int64

In [ ]:
# box_office.to_csv('cleaned_box_office_data.csv', index=False)

## Cleaning IMDB Data

In [120]:
imdb_df = pd.DataFrame(imdb)

In [121]:
imdb_genres_mode = imdb_df['genres'].mode()
print(imdb_genres_mode)

imdb_runtime_mean = round(imdb_df['runtime_minutes'].mean())
print(imdb_runtime_mean)

0    Drama
Name: genres, dtype: object
95


In [127]:
imdb_df['genres'] = imdb_df['genres'].fillna(imdb_genres_mode[0]).str.split(',').str[0]
imdb_df

,movie_id,genres,runtime_minutes,start_year,primary_title,numvotes,averagerating
0,tt0063540,Action,175.0,2013,Sunghursh,77,7.0
1,tt0066787,Biography,114.0,2019,One Day Before the Rainy Season,43,7.2
2,tt0069049,Drama,122.0,2018,The Other Side of the Wind,4517,6.9
3,tt0069204,Comedy,95.0,2018,Sabse Bada Sukh,13,6.1
4,tt0100275,Comedy,80.0,2017,The Wandering Soap Opera,119,6.5
...,...,...,...,...,...,...,...
73851,tt9913084,Documentary,75.0,2019,Diabolik sono io,6,6.2
73852,tt9914286,Drama,98.0,2019,Sokagin Çocuklari,136,8.7
73853,tt9914642,Documentary,95.0,2017,Albatross,8,8.5
73854,tt9914942,Drama,95.0,2019,La vida sense la Sara Amat,5,6.6


In [123]:
# Filling missing runtime values with mean
imdb_df['runtime_minutes'].fillna(imdb_runtime_mean, inplace=True)

In [128]:
imdb_df.nunique()

movie_id           73856
genres                25
runtime_minutes      289
start_year            10
primary_title      69993
numvotes            7349
averagerating         91
dtype: int64

In [129]:
imdb_df.isna().sum()

movie_id           0
genres             0
runtime_minutes    0
start_year         0
primary_title      0
numvotes           0
averagerating      0
dtype: int64

In [ ]:
# imdb_df.to_csv('cleaned_imdb_data.csv', index=False)